### PR TITLE
Destinations cdk: implement snapshotting spec test

### PR DESCRIPTION
**Pull request by edgao**
depends on https://github.com/airbytehq/airbyte/pull/45632 - CI will fail until that gets released

tl;dr add a test that:
* runs `spec` on the connector
* diffs that spec against the spec in `expected_spec.json`
* writes that spec back out to `expected_spec.json`

the original spec.json got deleted at some point in this stack? so I just created a new expected_spec.json - no strong opinion on where to put it, but I wanted it to be more obvious that it's not the "canonical" spec.

other things to note:
* junit isn't able to discover the test from inside cdk/testFixtures, hence the `class E2eSpecTest: SpecTest()`. I'm pretty sure there's a way to make this work but haven't bothered looking into it yet, and I think this is good enough for now.
* SpecTest actually writes out the new expected_spec.json for convenience. The expectation is that we'll just git commit that anytime we modify the spec.
* jackson's pretty printer doesn't behave the same as `prettier`, so I excluded the expected_spec.json files from airbyte-ci's format command.

I already committed the updated spec, but there's still a few diffs against the original spec. Pasting them here, I think they're all intentional (e.g. changing type, tweaking wording, swapping `example` -> `examples`)

```
The property "$.connectionSpecification.properties.test_destination.oneOf.0.properties.test_destination_type.const" in the expected json is not found
The property "$.connectionSpecification.properties.test_destination.oneOf.0.properties.logging_config.oneOf.1.properties.max_entry_count.description" didn't match. Expected "Max number of entries to log. This destination is for testing only. So it won't make sense to log infinitely. The maximum is 1,000 entries.", Received: "Number of entries to log. This destination is for testing only. So it won't make sense to log infinitely. The maximum is 1,000 entries."
The property "$.connectionSpecification.properties.test_destination.oneOf.0.properties.logging_config.oneOf.1.properties.nth_entry_to_log.example" in the expected json is not found
The property "$.connectionSpecification.properties.test_destination.oneOf.0.properties.logging_config.oneOf.1.properties.nth_entry_to_log.type" didn't match. Expected "number", Received: "integer"
The property "$.connectionSpecification.properties.test_destination.oneOf.0.properties.logging_config.oneOf.2.properties.max_entry_count.description" didn't match. Expected "Max number of entries to log. This destination is for testing only. So it won't make sense to log infinitely. The maximum is 1,000 entries.", Received: "Number of entries to log. This destination is for testing only. So it won't make sense to log infinitely. The maximum is 1,000 entries."
The property "$.connectionSpecification.properties.test_destination.oneOf.1.properties.test_destination_type.const" in the expected json is not found
The property "$.connectionSpecification.properties.test_destination.oneOf.2.properties.test_destination_type.const" in the expected json is not found
The property "$.connectionSpecification.properties.test_destination.oneOf.2.properties.millis_per_record.description" didn't match. Expected "Number of milli-second to pause in between records.", Received: "The number of milliseconds to wait between each record."
The property "$.connectionSpecification.properties.test_destination.oneOf.3.properties.test_destination_type.const" in the expected json is not found
```

---

example diff in expected_spec.json:
image

### CODE REVIEW

**Comment by edgao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/45468?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#45468** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/45468?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#45113** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/45113?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#44874** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/44874?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#45147** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/45147?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @edgao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-integrations/connectors/destination-e2e-test/src/main/kotlin/io/airbyte/integrations/destination/e2e_test/E2EDestinationSpecification.kt [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-
-@Singleton
-class E2EDestinationSpecification: DestinationSpecification {
-    override val supportedSyncModes = listOf(
-        DestinationSyncMode.OVERWRITE,
-        DestinationSyncMode.APPEND,
-        DestinationSyncMode.APPEND_DEDUP,
-    )
-}
+//package io.airbyte.integrations.destination.e2e_test
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-e2e-test/src/main/kotlin/io/airbyte/integrations/destination/e2e_test/E2EDestinationSpecification.kt:1]: build is broken, commented this out just for testing purposes

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * one, but after that, it should be pretty readable.
+ */
+open abstract class SpecTest :
+    IntegrationTest(
+        FakeDataDumper,
+        NoopDestinationCleaner,
+        NoopExpectedRecordMapper,
+    ) {
+    @Test
+    fun testSpec() {
```
Comment (postamar) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:46]: Can we please stop this pattern of defining junit tests in abstract classes? They're like a straightjacket and I strongly believe this pattern is responsible for the poor quality of the testing of the legacy connectors.

Instead, I encourage you to create test fixtures and define one-liner junit tests in the connectors.

Composition over inheritance blah blah blah...

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        assertEquals(
+            specMessages.size,
+            1,
+            "Expected to receive exactly one connection status message, but got ${specMessages.size}: $specMessages"
+        )
+
+        val spec = specMessages.first().spec
+        Files.write(
+            Path.of(EXPECTED_SPEC_FILENAME),
+            ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(spec).toByteArray()
```
Comment (postamar) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:62]:
```suggestion
            Jsons.writerWithDefaultPrettyPrinter().writeValueAsString(spec).toByteArray()
```

This has all the right jackson modules and settings and stuff.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        )
```
Comment (postamar) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:63]: This is a good idea and we should do more of that throughout the CDK.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+        val jsonMatcher: JsonMatcher =
+            CompositeJsonMatcher(
+                StrictJsonArrayPartialMatcher(),
+                LenientJsonObjectPartialMatcher(),
+                StrictPrimitivePartialMatcher(),
+            )
```
Comment (postamar) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:70]: FYI I'm not quite sure this combo of strict and lenient is the one we want. Just a heads up.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        val diff: JsonDiff = DiffGenerator.diff(expectedSpec, Jsons.serialize(spec), jsonMatcher)
```
Comment (postamar) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:71]: If you're writing to the file, perhaps you should also assert that the contents of the file haven't changed?

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        Assertions.assertEquals(
+            "",
+            OnlyErrorDiffViewer.from(diff).toString(),
+            "Spec snapshot test failed. Run this test locally and then `git diff <...>/expected_spec.json` to see what changed, and commit the diff if that change was intentional."
```
Comment (postamar) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:75]: yes! nice

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/consts.py [hunk 8]
```diff
@@ -41,6 +41,7 @@
     "airbyte-ci/connectors/metadata_service/lib/metadata_service/models/generated/**",  # These files are generated and should not be formatted
     "**/airbyte-ci/connectors/metadata_service/lib/tests/fixtures/**/invalid",  # This is a test directory with invalid and sometimes unformatted code
     "airbyte-ci/connectors/pipelines/tests/test_format/non_formatted_code",  # This is a test directory with badly formatted code
+    "airbyte-integrations/connectors/destination-**/expected_spec.json",  # These files are generated by tests using Jackson's pretty printer, and should not be reformatted.
```
Comment (postamar) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/consts.py:44]: kebab case instead of snake case? it's what prevails in the bulk cdk, might as well be consistent

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt (hunk 2 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:46]: (tl;dr imo the abstract class thing is the least-shitty pattern available)

hm, are you suggesting something like this? (also can you say more about what goes wrong with the abstract class pattern? I personally think it's... sufficiently not terrible, though obviously it's pretty sucky)
```kotlin
package cdk.blah
class DestinationTests {
  fun testSpec() {...}
  fun testWrite() {...}
}

package destination-blah
class DestinationBlahSpecTest {
  @Test
  fun testSpec() { DestinationTests.testSpec() }

  @Test
  fun testWrite() { DestinationTests.testWrite() }
}
```

if I'm reading you correctly, then I disagree. Specifically: I think these two requirements are nonnegotiable:
* test cases must be opt-out (i.e. if we add a new test case in CDK v1.2.3, then upgrading destination-foo to cdk 1.2.3 should automatically activate that test for destination-foo)
    * (reasoning: nobody is going to read through the cdk changelog and remember to opt into new tests, every time they bump the cdk version. And destinations want to put most of our tests in the CDK.)
* it must be easy to run an individual test case from intellij (it sucks that this is a real problem, but de facto we all use intellij, and it imposes some annoying constraints on the problem - e.g. we can't use DynamicTest, parameterized tests are a little annoying, etc).
    * (IMO it's self-evident that this is important)

and requiring destination authors to write those `fun testFoo() { Base.testFoo() }` methods fails the first criterion

Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:46]: haven't done any research myself (obviously kotest looms large, with our shift to kotlin, but I know nothing about the pros/cons). I think the problem is more with intellij than the test framework though (with 0 evidence)

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt (hunk 5 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:70]: it worked well enough to catch some legitimate diffs against the original destination-e2e-test's spec, so :shrug: presumably good enough for now

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/consts.py (hunk 8 above)
Comment (edgao) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/consts.py:44]: makes sense. I'm actually going to move this to a separate PR, pretty sure the airbyte-ci change can't be in the same PR as expected-spec.json

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt (hunk 2 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt:47]: also, I have a note to myself - this test (unlike most of our other tests) actually _could_ live purely in the CDK as a concrete class. Just need to get gradle/junit to search the full classpath, instead of just the destination-foo jar. Currently `gw :air-int:conn:dest-e2e-test:integrationTestJava --tests io.airbyte.cdk.test.spec.SpecTest` errors b/c no tests are discovered, but that feels solvable.

(or something along those lines, I didn't really dig into it)

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt (hunk 4 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt]: credit goes to @stephane-airbyte for this one :D

**Review by johnny-schmidt — APPROVED** [review]
